pandas/core/tools/datetimes.py

```diff
@@ -60,7 +74,42 @@ def _maybe_cache(arg, format, cache, convert_listlike):
     return cache_array
 
 
-def _convert_and_box_cache(arg, cache_array, box, errors, name=None):
+def _box_as_indexlike(
+    dt_array: ArrayLike,
+    tz=None,
+    name: str = None
+) -> Union[ABCIndex, ABCDatetimeIndex]:
+    """
+    Properly boxes the ndarray of datetimes to DatetimeIndex
+    if it is possible or to generic Index instead
+
+    Parameters
+    ----------
+    dt_array: 1-d array
+        array of datetimes to be boxed
+    tz : object
+        None or 'utc'
+    name : string, default None
+        Name for a resulting index
+
+    Returns
+    -------
+    result : datetime of converted dates
+        - DatetimeIndex if convertible to sole datetime64 type
+        - general Index otherwise
+    """
+    from pandas import DatetimeIndex, Index
+    if is_datetime64_dtype(dt_array):
+        return DatetimeIndex(dt_array, tz=tz, name=name)
+    return Index(dt_array, name=name)
```

Comment:
We need to apply `tz` to this result if `Index(dt_array, name=name)` returns a `DatetimeIndex`

Reply:
@jreback @mroeschke After a close look at the Index constructor(`pandas/core/indexes/base.py`), I had a question about `dtype` default value. The documentation says that it should be `object`, but actually `None` is used. What case is correct?

Reply:
its default is ``None``, however it will attempt to infer the type of thing passed, so you *may* need to explicity pass ``dtype='object'`` to force it to object here.